File: fn/src/test/java/org/sqlbuilder/fn/TestListCollector.java
package org.sqlbuilder.fn;

import org.junit.Test;

public class TestListCollector
{
	static class W<T> implements HasId, SetId
	{
		private int id;

		final T wrapped;

		W(T wrapped)
		{
			this.wrapped = wrapped;
		}

		static <T> W<T> w(T wrapped)
		{
			return new W<>(wrapped);
		}

		@Override
		public String toString()
		{
			return wrapped.toString();
		}

		@Override
		public Integer getIntId()
		{
			return id;
		}

		@Override
		public void setId(final int id)
		{
			this.id = id;
		}
	}

	@Test
	public void testListCollector()
	{
		ListCollector<W<String>> collector = new ListCollector<>();
		collector.add(W.w("one"));
		collector.add(W.w("two"));
		collector.add(W.w("two"));
		collector.add(W.w("one"));
		collector.add(W.w("three"));
		for (var item : collector)
		{
			// System.out.printf("%s%n", item);
			// System.out.printf("%s %d%n", item, collector.indexOf(item) + 1);
			System.out.printf("%s %d%n", item, item.getIntId());
		}
	}
}

File: fn/src/test/java/org/sqlbuilder/fn/TestListCollector.kt
package org.sqlbuilder.fn

import org.junit.Test
import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.ListCollector
import org.sqlbuilder.common.SetId

class TestListCollector {
    internal class W<T>(val wrapped: T?) : HasId, SetId {

        private var id = 0

        override fun toString(): String {
            return wrapped.toString()
        }

        override fun getIntId(): Int {
            return id
        }

        override fun setId(id: Int) {
            this.id = id
        }

        companion object {

            fun <T> w(wrapped: T?): W<T?> {
                return W<T?>(wrapped)
            }
        }
    }

    @Test
    fun testListCollector() {
        val collector = ListCollector<W<String?>>()
        collector.add(W.Companion.w<String?>("one"))
        collector.add(W.Companion.w<String?>("two"))
        collector.add(W.Companion.w<String?>("two"))
        collector.add(W.Companion.w<String?>("one"))
        collector.add(W.Companion.w<String?>("three"))
        for (item in collector) {
            // println(item)
            // println("$item ${collector.indexOf(item) + 1}")
            println("$item ${item.intId}")
        }
    }
}
